File: backend/src/services/SystemConfigService.cpp
#include "services/SystemConfigService.h"
#include "repositories/MockSystemConfigRepository.h"
#include "services/OperationLogService.h"
#include "services/PermissionService.h"
#include "utils/ApiResponse.h"
#include "utils/ErrorCode.h"
#include "utils/JwtUtil.h"
#include "utils/TimeUtil.h"

namespace smart_home {

static std::string configJson(const SystemConfig& c) {
    return "{\"configId\":" + std::to_string(c.configId) + ",\"homeId\":" + std::to_string(c.homeId) +
           ",\"configKey\":" + ApiResponse::quote(c.configKey) + ",\"configValue\":" + ApiResponse::quote(c.configValue) +
           ",\"configType\":" + ApiResponse::quote(c.configType) + ",\"description\":" + ApiResponse::quote(c.description) + "}";
}

std::string SystemConfigService::listConfig(const std::string& token, long long homeId) {
    const auto payload = JwtUtil::parseToken(token);
    if (!payload) return ApiResponse::error(UNAUTHORIZED, "unauthorized");
    if (!PermissionService().isMemberOfHome(payload->userId, homeId)) return ApiResponse::error(FORBIDDEN, "permission denied");
    MockSystemConfigRepository repo;
    const auto configs = repo.findByHome(homeId);
    std::string data = "[";
    for (size_t i = 0; i < configs.size(); ++i) {
        if (i) data += ",";
        data += configJson(configs[i]);
    }
    data += "]";
    return ApiResponse::success(data);
}

std::string SystemConfigService::updateConfig(const std::string& token, long long homeId, const std::string& key,
                                              const std::string& value, const std::string& description) {
    const auto payload = JwtUtil::parseToken(token);
    if (!payload) return ApiResponse::error(UNAUTHORIZED, "unauthorized");
    if (!PermissionService().canUpdateSystemConfig(payload->userId, homeId)) return ApiResponse::error(FORBIDDEN, "permission denied");
    MockSystemConfigRepository repo;
    const SystemConfig saved = repo.upsert(SystemConfig{0, homeId, key, value, "ALARM", description, payload->userId, TimeUtil::now()});
    OperationLogService().write(payload->userId, homeId, "UPDATE_SYSTEM_CONFIG", "update system config", "SYSTEM_CONFIG", saved.configId);
    return ApiResponse::success(configJson(saved));
}

} // namespace smart_home

File: smart_home/smart home/backend/app/services/system_config_service.py
from app.core.permissions import Permission
from app.repositories.factory import get_system_config_repository
from app.services.account_operation_log_service import AccountOperationLogService
from app.services.permission_service import PermissionService


class SystemConfigService:
    def __init__(self):
        self.configs = get_system_config_repository()
        self.permissions = PermissionService()
        self.logs = AccountOperationLogService()

    def get_config(self, home_id: int, user_id: int) -> dict:
        self.permissions.require_home_permission(home_id, user_id, Permission.HOME_VIEW)
        return self.configs.get_by_home(home_id).__dict__

    def update_config(self, home_id: int, user_id: int, **kwargs) -> dict:
        self.permissions.require_home_permission(home_id, user_id, Permission.SYSTEM_CONFIG)
        config = self.configs.update(home_id, **kwargs)
        self.logs.write(user_id, home_id, "UPDATE_SYSTEM_CONFIG", "update system config", "SYSTEM_CONFIG", home_id)
        return config.__dict__

    def list_logs(self, home_id: int, user_id: int) -> list[dict]:
        self.permissions.require_home_permission(home_id, user_id, Permission.LOG_VIEW)
        return self.logs.list_by_home(home_id)
